trl/trainer/online_dpo_trainer.py

```diff
@@ -326,8 +356,11 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
         del output, logits, all_logprobs  # free memory
 
         # Same for the reference model
-        with torch.no_grad():
+        if self.ref_model is not None:
             ref_output = self.ref_model(prompt_completion_ids, attention_mask=prompt_completion_mask)
```

Comment:
is `with torch.no_grad():` needed here?

Reply:
In fact, I would wrap lines 358 -> 373 in the same no grad context. As all the ref model and reward model stuff does not require grads.

Reply:
I don't think it's needed because `self.ref_model = create_reference_model(ref_model)` disable the gradients

Reply:
New:

```diff
  if ref_model is None:  # No ref model provided, the most common case
      ...
  else:  # rare case, the user provided a ref model
-     self.ref_model = create_reference_model(ref_model)  # copy, disable gradients, set eval mode
+     self.ref_model = ref_model
+     self.ref_model.eval()
```

and global `torch.no_grad` context, as suggested